File: RB110/small_problems/medium1/diamonds.rb
=begin

Write a method that displays a 4-pointed diamond in an n x n grid, where n is an odd 
integer that is supplied as an argument to the method. You may assume that the argument 
will always be an odd integer.

Input:
  -An integer

Output:
  -Display a four pointed diamond in a n x n grid

Rules:
  Explicit requirments:
    -intput integer will always be a n odd number
  Implicit:
    -Display the diamond as a '*' symbol

  Ouestions?
    -Are there whitespace above and below the grid?

Test cases / Exsamples:

diamond(1)

*

diamond(3)

 *
***
 *

 diamond(9)

    *
   ***
  *****
 *******
*********
 *******
  *****
   ***
    *

data structure:
  -integers
  -arrays

algorithm:
  -create a grid that is input integer by input integer
  -Save a '*' (star)
  -Save a ' ' (whitespace)
  -create a print array
    -Save pattern starting with 1 star surrounded by white spaces
  -loop n number of times
    -puts print array
    -change print array by add 2 stars and substracting 2 white spaces
    -when print array as only stars reverse the process
  -end loop

=end

def star_line(star, space)
  line = (' ' * space) + ('*' * star) + (' ' * space)
  line
end

def diamond(num)
  star = 1
  space = num / 2
  top_half = space + 1
  bottom_half = space

  top_half.times do
    puts star_line(star, space)
    star += 2
    space -= 1
  end

  star -= 2
  space += 1

  4.times do
    star -= 2
    space += 1
    puts star_line(star, space)
  end
end

diamond(9)

File: LS215/exerises/problem_solving/diamonds.js
/*
Write a function that displays a four-pointed diamond in an nxn grid, where n is an odd integer supplied as 
an argument to the function. You may assume that the argument will always be an odd integer.

Input:
  - A odd number

Output
  - A four pointed diamond on a n x n grid

Rules:
  - Assume that the input integer is odd
  - The diamond should be represented with stars '*' and whitespaces
  - Each row will have astericks
    - 5 rows will have 1, 3, 5, 3, 1 stars
    - 5 rows will have 2, 1, 0, 1, 2 white spaces

Questions:
  - Will the input every be a negative integer? What should i do?
  - Will the input be a data type other then an integer?
  - Will there be more then one argument? No arguments?
  - Will the input integer be larger number?

Data Structure:
  - 2 dimensional array for the map the grid
  - Strings as the elements in the grid
  - input is an integer

Algorithm:
  - Create the algorithm for astericks 1, 3, ..n n-2 ...
  - Create the algorithm for white spaces (n - astericks) / 2




*/

function createAsterisks(num) {
  let array = [];

  for (let i = 1; i <= num; i += 2) {
    array.push(i);
  }
  for (let j = (num - 2); j > 0; j -= 2) {
    array.push(j);
  }

  return array;
}

function createWhiteSpaces(asterisks, num) {
  let array = [];

  asterisks.forEach(asterisk => {
    array.push((num - asterisk) / 2);
  });

  return array;
}


function diamond(num) {
  let asterisks = createAsterisks(num);
  let whiteSpaces = createWhiteSpaces(asterisks, num);

  for (let i = 0; i <= num; i++) {
    let space = ' '.repeat(whiteSpaces[i]);
    let ast = '*'.repeat(asterisks[i]);
    console.log(space + ast + space);
  }
}



// diamond(1);
// logs
// *

// diamond(3);
// logs
//  *
// ***
//  *
// diamond(5)

// diamond(9);
// logs
//     *
//    ***
//   *****
//  *******
// *********
//  *******
//   *****
//    ***
//     *

diamond(15)